File: backend/src/main/scala/domains/contest/api/EvaluateContestAccess.scala
package domains.contest.api

import cats.effect.IO
import domains.auth.api.InternalOnlyAuthenticatedApi
import domains.auth.objects.internal.AuthenticatedUser
import domains.contest.objects.{ContestId, ContestSlug, ContestTitle}
import domains.contest.table.contest.ContestTable
import domains.contest.utils.ContestAccessRules
import domains.problem.objects.ProblemId
import domains.usergroup.api.ListUserGroupSlugsForMember
import org.http4s.Method
import shared.api.ApiPath

import java.sql.Connection
import java.time.Instant

object EvaluateContestAccess extends InternalOnlyAuthenticatedApi[EvaluateContestAccess.Input, Option[EvaluateContestAccess.Result]]:

  final case class Input(
    contestSlug: ContestSlug,
    problemId: Option[ProblemId]
  )

  final case class Result(
    contestId: ContestId,
    contestSlug: ContestSlug,
    contestTitle: ContestTitle,
    contestStarted: Boolean,
    contestEnded: Boolean,
    isRegistered: Boolean,
    containsProblem: Boolean,
    canViewContest: Boolean,
    canViewContestDetail: Boolean,
    canManageContest: Boolean,
    canViewLinkedContestProblem: Boolean,
    canManageLinkedContestProblem: Boolean,
    canSubmitContestProblem: Boolean
  )

  override val method: Method = Method.POST
  override val path: ApiPath = ApiPath("/api/internal/contests/evaluate-access")

  override def plan(connection: Connection, actor: AuthenticatedUser, input: Input): IO[Option[Result]] =
    ContestTable.findBySlug(connection, input.contestSlug).flatMap {
      case None =>
        IO.pure(None)
      case Some(contest) =>
        for
          actorGroupSlugs <- ListUserGroupSlugsForMember.plan(connection, actor.username)
          isRegistered <- ContestTable.isRegistered(connection, contest.id, actor.username)
          now = Instant.now()
          actorGroupSlugSet = actorGroupSlugs.slugs.toSet
          containsProblem = input.problemId.exists(problemId => contest.problems.exists(_.id.value == problemId.value))
          canViewContest = ContestAccessRules.canViewContest(actor, contest, actorGroupSlugSet)
          canViewContestDetail = ContestAccessRules.canViewContestDetail(actor, contest, actorGroupSlugSet, isRegistered, now)
          canManageContest = ContestAccessRules.canManageContest(actor, contest, actorGroupSlugSet)
          canViewLinkedContestProblem =
            ContestAccessRules.canViewLinkedContestProblem(actor, contest, actorGroupSlugSet, isRegistered, containsProblem, now)
          canManageLinkedContestProblem =
            ContestAccessRules.canManageLinkedContestProblem(actor, contest, actorGroupSlugSet, containsProblem)
          canSubmitContestProblem =
            ContestAccessRules.canSubmitContestProblem(contest, isRegistered, containsProblem, now)
        yield Some(
          Result(
            contestId = contest.id,
            contestSlug = contest.slug,
            contestTitle = contest.title,
            contestStarted = !now.isBefore(contest.startAt),
            contestEnded = now.isAfter(contest.endAt),
            isRegistered = isRegistered,
            containsProblem = containsProblem,
            canViewContest = canViewContest,
            canViewContestDetail = canViewContestDetail,
            canManageContest = canManageContest,
            canViewLinkedContestProblem = canViewLinkedContestProblem,
            canManageLinkedContestProblem = canManageLinkedContestProblem,
            canSubmitContestProblem = canSubmitContestProblem
          )
        )
    }

File: frontend/src/apis/contest/EvaluateContestAccess.ts
import type { APIWithSessionMessage } from '@/system/api/api-message'
import type { ContestSlug } from '@/objects/contest/ContestSlug'
import type { ProblemId } from '@/objects/problem/ProblemId'

/** 内部比赛访问评估请求体；problemId 为空表示只评估比赛本身访问。 */
type EvaluateContestAccessBody = {
  contestSlug: ContestSlug
  problemId: ProblemId | null
}

/** 评估当前会话对比赛/比赛题目的访问；输出类型尚未建模，权限由后端判断。 */
/** FIXME-CN: 后端返回 Option[EvaluateContestAccess.Result]，这里用 unknown 会让调用方失去权限字段类型保护，应补前端镜像类型。 */
export class EvaluateContestAccess implements APIWithSessionMessage<unknown> {
  declare readonly responseType?: unknown
  readonly method = 'POST'
  readonly apiPath = 'internal/contests/evaluate-access'
  private readonly contestSlug: ContestSlug
  private readonly problemId: ProblemId | null

  constructor(contestSlug: ContestSlug, problemId: ProblemId | null) {
    this.contestSlug = contestSlug
    this.problemId = problemId
  }

  body(): EvaluateContestAccessBody {
    return {
      contestSlug: this.contestSlug,
      problemId: this.problemId,
    }
  }
}
